lib/mcp/app/resource_template.rb

```diff
@@ -0,0 +1,142 @@
+        def mime_type(value)
+          @mime_type = value
+        end
+        # standard:enable Lint/DuplicateMethods,Style/TrivialAccessors
+
+        def call(&block)
+          @handler = block
+        end
+
+        def to_resource_template_hash
+          raise ArgumentError, "Name must be provided" if @name.empty?
+
+          {
+            uri_template: @uri_template,
+            name: @name,
+            mime_type: @mime_type,
+            description: @description,
+            handler: @handler,
+            variables: @variables
+          }
+        end
+        
+        # Extract variables from a URI template
+        # e.g., "channels://{channel_id}" => ["channel_id"]
+        def extract_variables(uri_template)
+          variables = []
+          uri_template.scan(/\{([^}]+)\}/) do |match|
+            variables << match[0]&.to_sym
+          end
+          variables
+        end
+        
+        # Creates a pattern for matching URIs against this template
+        def to_pattern
+          pattern_string = Regexp.escape(@uri_template).gsub(/\\\{[^}]+\\\}/) do |match|
+            var_name = match.gsub(/\\\{|\\\}/, '')
+            "([^/]+)"
+          end
+          Regexp.new("^#{pattern_string}$")
+        end
+        
+        # Extract variable values from a concrete URI based on the template
+        # e.g., template: "channels://{channel_id}", uri: "channels://123" => {"channel_id" => "123"}
+        def extract_variable_values(uri)
+          pattern = to_pattern
+          match = pattern.match(uri)
+          return {} unless match
+          
+          result = {}
+          @variables.each_with_index do |var_name, index|
+            result[var_name] = match[index + 1]
+          end
+          result
+        end
+      end
+
+      def register_resource_template(uri_template, &block)
+        builder = ResourceTemplateBuilder.new(uri_template)
+        builder.instance_eval(&block)
+        template_hash = builder.to_resource_template_hash
+        resource_templates[uri_template] = template_hash
+        template_hash
+      end
+      
+      # Find a template that matches the given URI and extract variable values
+      def find_matching_template(uri)
+        resource_templates.each do |template_uri, template|
+          builder = ResourceTemplateBuilder.new(template_uri)
+          variable_values = builder.extract_variable_values(uri)
+          return [template, variable_values] unless variable_values.empty?
+        end
+        [nil, {}]
+      end
+
+      def list_resource_templates(cursor: nil, page_size: nil)
+        start_index = cursor&.to_i || 0
+        values = resource_templates.values
+
+        if page_size.nil?
+          paginated = values[start_index..]
+          next_cursor = ""
+        else
+          paginated = values[start_index, page_size]
+          has_next = start_index + page_size < values.length
+          next_cursor = has_next ? (start_index + page_size).to_s : ""
```

Comment:
I think it would be good to align this with `list_tools` and set it to `nil` as well.

https://github.com/modelcontextprotocol/specification/blob/6828f3ef6300b25dd2aaff2a2e5e81188bdbd22e/schema/2024-11-05/schema.ts#L332-L334